### ISSUE TITLE
Optimize volumemanager run loop

### ISSUE THREAD

**Issue by linxiulei** [kubernetes/kubernetes#126049]
### What would you like to be added?

Change how volume management tasks are scheduled to execute to reduce its overhead.

### Why is this needed?

There are two parts in volume management:

* populating the desired state of volumes
* reconciling the desired state and actual state

Both parts are scheduled to run for every [100ms](https://github.com/kubernetes/kubernetes/blob/7ec344dcef9a04920578aabe4e122bc0b3404afe/pkg/kubelet/volumemanager/volume_manager.go#L58-L62), which might be unnecessary for nodes with very steady state (e.g. standalone kubelet). 

Two optimization options:

1. Just making the run interval configurable so the users can decide accordingly. However, this not only increases the configuration complexity but also increases the latency to reconcile the volumes
2. Making the populating first watch over pod changes, which can be a signal passed from kubelet PLEG/housekeeping so populating is only triggered when there are changes in pods. Then Also making reconciling watch over signals passed by populating, which only triggers a signal when there are differences populated. This optimization could be more complicated with trivial retries to ensure the reconciliation is successfully executed.

With optimization, I expect 1%~1.5% CPU per node reduced

**Comment by bouaouda-achraf** [comment]
hello 
I can work on this issue feature ? 
/assign

**Comment by SergeyKanzhelev** [comment]
If this is primarily targeting the standalone mode - should we make a special case defaults for it?

I don't think exposing two interval durations with the cryptic names is a proper way to go here. It may be way better to find a condition that will let kubelet be more efficient and not run those that often.

**Comment by bouaouda-achraf** [comment]
Hello 👋
Please correct me if I’m wrong 🙏:

By design, the kubelet  volume manager runs based on asynchronous **loop**. I preferred exposing these variables because it will not disrupt the existing architecture. Moving to an **event-based** design seems costly at this initial stage.

Proposal:

	•	Add a new KEP (Kubernetes Enhancement Proposal) to transition the volume manager to an event-based design.
	•	Accept the existing implementation to address current needs.

Waiting for your feedback. 🙏

**Comment by googs1025** [comment]
I have a bit of a concern: If we only check if it's greater than 0, I'm not sure if there might be any issues on the system level. If a user enters a very large number or sets this number to 1m, would it affect the kubelet itself? Because this is a value that users are not aware of, and of course, users would prefer it to be as small as possible. But I'm not sure if it would impact the functionality of the kubelet.

**Comment by bouaouda-achraf** [comment]
Hello @googs1025 
I replied to you on the pr comment .

**Comment by SergeyKanzhelev** [comment]
> Proposal:
> 
> ```
> •	Add a new KEP (Kubernetes Enhancement Proposal) to transition the volume manager to an event-based design.
> •	Accept the existing implementation to address current needs.
> ```
> 
> Waiting for your feedback. 🙏

Generally sounds like a good idea. Perhaps you can craft a small writeup.

As a side note, if there is a simpler version that will just increase the reconciliation timeout if there were no Pod changes (instead of a config change), that may also be a good design.

**Comment by bouaouda-achraf** [comment]
Ok @SergeyKanzhelev 👍 I will write the kep . 
« if there is a simpler version  « your are referring to the existing pr or the kep ?

**Comment by SergeyKanzhelev** [comment]
No, I am suggesting that the KEP or feature may go in steps, and the first step may be to react on events differently - change the timer instead of removing the loop completely

**Comment by bouaouda-achraf** [comment]
@SergeyKanzhelev a last confirmation please : 
should I remove the ability for a user to customise the configuration  ?

**Comment by SergeyKanzhelev** [comment]
> @SergeyKanzhelev a last confirmation please : should I remove the ability for a user to customise the configuration ?

Yes, unless somebody from sig storage thinks it is a good idea, but I don't think so.

**Comment by bouaouda-achraf** [comment]
> > @SergeyKanzhelev a last confirmation please : should I remove the ability for a user to customise the configuration ?
> 
> Yes, unless somebody from sig storage thinks it is a good idea, but I don't think so.

Thank you , it’s clear now 🙏 .

**Comment by bouaouda-achraf** [comment]
@SergeyKanzhelev 

[SIG STORAGE meeting 19/09/2024](https://docs.google.com/document/d/1-8KEG8AjAgKznS9NFm3qWqkGyCHmvU6HVl0sk5hwoAE/edit) : it's OK for DSWP to move to event-based approach but for the reconciler, it is not limited to CRI event. 
I will start with DWSP for the first step, and after that, I will check the reconciler logic.
